### PR TITLE
[Driver][SYCL]Emit an error if c compilation is forced using -x c or …

### PR DESCRIPTION
**Pull request by hchilama**
…-x c-header when -fsycl mode is used

Signed-off-by: hchilama <harini.chilamantula@intel.com>

### CODE REVIEW

File: clang/lib/Driver/Driver.cpp [hunk 1]
```diff
@@ -4057,6 +4057,13 @@ class OffloadingActionBuilder final {
         SYCLTripleList.push_back(
             C.getDriver().MakeSYCLDeviceTriple(SYCLTargetArch));
       }
+      // Emit an error if c-compilation is forced in -fsycl mode
+      if (HasValidSYCLRuntime && Args.hasArg(options::OPT_x))
+        for (StringRef XValue : Args.getAllArgValues(options::OPT_x)) {
+          if (XValue == "c" || XValue == "c-header")
+            C.getDriver().Diag(clang::diag::err_drv_fsycl_with_c_type)
+                << "-x " << XValue;
+        }
```
Comment (AGindinson) [inline clang/lib/Driver/Driver.cpp]: Should this be moved to `Driver::CreateOffloadingDeviceToolchains()`? It seems that doing such a diagnostic would make sense before any offloading actions' construction is attempted.

File: clang/include/clang/Basic/DiagnosticDriverKinds.td [hunk 2]
```diff
@@ -255,6 +255,8 @@ def err_drv_expecting_fopenmp_with_fopenmp_targets : Error<
   "The option -fopenmp-targets must be used in conjunction with a -fopenmp option compatible with offloading, please use -fopenmp=libomp or -fopenmp=libiomp5.">;
 def err_drv_expecting_fsycl_with_sycl_opt : Error<
   "The option %0 must be used in conjunction with -fsycl to enable offloading.">;
+def err_drv_fsycl_with_c_type : Error<
+  "The option %0%1 must not be used in conjunction with -fsycl which expects C++ source.">;
```
Comment (AGindinson) [inline clang/include/clang/Basic/DiagnosticDriverKinds.td]: How would the driver behave on simply receiving a c source/header without `-x c++` for SYCL compilation?
Unless this scenario is already covered, it seems that a more general diagnostic would only mention the input type, and be triggered in both the `.c` and `-x c` cases. Aka:
`The %0 file type must not be compiled with -fsycl which expects C++ source.`

File: clang/lib/Driver/Driver.cpp (hunk 1 above)
Comment (hchilama) [reply clang/lib/Driver/Driver.cpp]: Sure, I will try moving there and test it.

File: clang/include/clang/Basic/DiagnosticDriverKinds.td (hunk 2 above)
Comment (hchilama) [reply clang/include/clang/Basic/DiagnosticDriverKinds.td]: I tested it already by passing a c/cheader without -x option to clang++, it doesn't force c compilation on the c/c-header based on the type of the file.  Instead I got the same output as if it is cpp source file.

Comment (hchilama) [reply clang/include/clang/Basic/DiagnosticDriverKinds.td]: Here is output from the tool that suggests the .c file is compiled as if it is .cpp. So, the behavior of the tool when -x(c/c-header) is as expected. I think this diagnostic should only address user specifically forcing c-compilation.  

clang-11: warning: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated [-Wdeprecated]

**Comment by AGindinson** [comment]
@hchilama Harini, when updating the patch, could you please also update the signed-off field to include your full name? This is in accordance with https://github.com/intel/llvm/blob/sycl/CONTRIBUTING.md.

File: clang/lib/Driver/Driver.cpp [hunk 3]
```diff
@@ -769,6 +769,15 @@ void Driver::CreateOffloadingDeviceToolChains(Compilation &C,
     }
     return SYCLArg;
   };
+
+  // Emit an error if c-compilation is forced in -fsycl mode
+  if (HasValidSYCLRuntime && C.getInputArgs().hasArg(options::OPT_x))
+    for (StringRef XValue : C.getInputArgs().getAllArgValues(options::OPT_x)) {
```
Comment (AGindinson) [inline clang/lib/Driver/Driver.cpp:775]: This seems sufficient:
```suggestion
  if (HasValidSYCLRuntime)
    for (StringRef XValue : C.getInputArgs().getAllArgValues(options::OPT_x)) {
```

Comment (hchilama) [reply clang/lib/Driver/Driver.cpp:775]: Yes I thought so too but in majority cases we don't have -x options so thought this offers better performance. I can make suggestion as you pointed if the performance drop expectation is negligible.

Comment (AGindinson) [reply clang/lib/Driver/Driver.cpp:775]: Actually, I don't see why an extra iteration over the arguments' list would bring better performance. We'll still have to traverse the full collection once, and the extra check itself requires such a traversal just like all values extraction.
Another thing that's occurred to me right now: would it make sense to get only the last argument value, which is supposed to override the preceding ones?
```suggestion
  if (HasValidSYCLRuntime)
    StringRef XValue = C.getInputArgs().getLastArg(options::OPT_x);
    if (!XValue.empty()) {
```

Comment (AGindinson) [reply clang/lib/Driver/Driver.cpp]: Fine with me if @mdtoguchi agrees

Comment (hchilama) [reply clang/lib/Driver/Driver.cpp]: Thank you AGindinson for the approval. I will ask Mike Toguchi for the review

**Pull request merged**